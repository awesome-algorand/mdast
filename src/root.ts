/**
 * @packageDocumentation
 * @document ./root.guide.md
 */

import type { Root } from 'mdast';
import { fromMarkdown } from 'mdast-util-from-markdown';
import { AwesomeContent, toRoots } from './content.js';

/**
 * Represents the AwesomeRoot type, which extends the Root type with additional optional properties.
 */
export type AwesomeRoot = Root & {
  /**
   * Represents the children elements contained within a component or structure.
   * Each child is an instance of the AwesomeContent type, which defines the
   * specific structure and properties of each child.
   *
   * This variable is optional and may be undefined if no child elements are present.
   */
  children?: AwesomeContent[];
  /**
   * An optional object that represents data with an optional category field.
   */
  data?: {
    /**
     * Represents the category or classification of an item or entity.
     * This property is optional and may not always be defined.
     */
    category?: string;
  };
};

/**
 * Converts a **Markdown** string into an array of {@link AwesomeRoot} nodes
 * by processing it and trimming a specified number of elements.
 *
 * @param {string} markdown - The Markdown content to process.
 * @param {number} [trim=5] - The number of elements to trim from the parsed Markdown structure.
 * @return {Root[]} An array of root nodes created from the processed Markdown content.
 * @see {@link object.fromRoot | fromRoot}
 * @see {@link object.fromRoots | fromRoots}
 */
export function fromString(markdown: string, trim: number = 0): AwesomeRoot[] {
  return toRoots(fromMarkdown(markdown).children.slice(trim));
}
