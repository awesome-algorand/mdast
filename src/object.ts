/**
 * @packageDocumentation
 * @document ./object.guide.md
 */
import { select } from 'unist-util-select';
import { toRoots } from './content.js';
import { AwesomeLink, fromList } from './links.js';
import { AwesomeRoot } from './root.js';

/**
 * Represents an AwesomeObject, which contains a title, optional description, subcategories, and links.
 */
export type AwesomeObject = {
  /**
   * The title of the object, derived from the `heading`
   */
  title: string;
  /**
   * Represents an optional description, derived from the `blockquote`
   */
  description?: string;
  /**
   * An optional property representing an array of AwesomeObject instances.
   * Each element in the array corresponds to a `heading` at a lower depth than the current.
   */
  subcategories?: AwesomeObject[];
  /**
   * Represents an optional array of AwesomeLink objects.
   * Each link in the array corresponds to a `list` of `url` and `text` items.
   *
   * @type {AwesomeLink[] | undefined}
   */
  links?: AwesomeLink[];
};

/**
 * Converts an array of AwesomeRoot objects into a mapped array using the fromRoot transformation.
 *
 * @param {AwesomeRoot[]} roots - Array of AwesomeRoot objects to be transformed.
 * @return {any[]} A new array resulting from applying the fromRoot transformation to each AwesomeRoot object.
 */
export function fromRoots(roots: AwesomeRoot[]) {
  return roots.map((x) => fromRoot(x));
}

/**
 * Converts an AwesomeRoot object into an AwesomeObject by extracting and transforming its properties.
 *
 * @param {AwesomeRoot} root - The root object containing structured data to be transformed.
 * @return {AwesomeObject} - The transformed object containing title, description, links, and subcategories (if available).
 */
export function fromRoot(root: AwesomeRoot): AwesomeObject {
  const obj: AwesomeObject = {
    // @ts-expect-error we know it is a text value
    title: select('text', root.children[0])!.value,
  };

  // Remove the first heading
  let children = root.children.slice(1);

  if (children.length && children[0].type === 'blockquote') {
    // @ts-expect-error we know it is a text value
    obj.description = select('text', root.children[1])!.value;
    children = children.slice(1);
  }

  if (children.length && children[0].type === 'list') {
    obj.links = fromList(children[0]);
    children = children.slice(1);
  }

  if (children.length && children[0].type === 'heading') {
    obj.subcategories = toRoots(children, children[0].depth).map((x) =>
      fromRoot(x),
    );
  }

  return obj;
}
