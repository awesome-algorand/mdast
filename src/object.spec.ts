import { describe, it, expect, beforeEach } from 'vitest';
import README from './__fixtures__/SPEC.md?raw';
// #region import
import { fromString } from '@awesome-algorand/mdast/root';
import { fromRoots } from '@awesome-algorand/mdast/object';
const markdown = `
## Category with List

- [Item](https://example.com) - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque posuere aliquam semper. Nulla facilisis mollis dui ac aliquam.
- [Item](https://example.com) - Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque posuere aliquam semper. Nulla facilisis mollis dui ac aliquam.
`;
let roots = fromString(markdown);
// #endregion import

describe('@awesome-algorand/mdast/object', () => {
  beforeEach(() => {
    roots = fromString(README);
  });
  it('(OK) snapshot', () => {
    // #region example
    // Map all Roots into AwesomeObjects
    const objs = fromRoots(roots);
    // #endregion example
    expect(objs).toMatchSnapshot();
  });
});
