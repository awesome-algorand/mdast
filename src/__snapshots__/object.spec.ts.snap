// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@awesome-algorand/mdast/object > (OK) snapshot 1`] = `
[
  {
    "links": [
      {
        "description": "Item",
        "url": "https://example.com",
      },
      {
        "description": "Item",
        "url": "https://example.com",
      },
    ],
    "title": "Category with List",
  },
  {
    "description": "Sed sit amet sollicitudin dolor. Donec scelerisque sed tortor sed elementum.",
    "links": [
      {
        "description": "Item",
        "url": "https://example.com",
      },
      {
        "description": "Item",
        "url": "https://example.com",
      },
    ],
    "title": "Category with Block Quote and List",
  },
  {
    "subcategories": [
      {
        "links": [
          {
            "description": "Item",
            "url": "https://example.com",
          },
          {
            "description": "Item",
            "url": "https://example.com",
          },
        ],
        "title": "SubCategory with List",
      },
      {
        "description": "Sed sit amet sollicitudin dolor. Donec scelerisque sed tortor sed elementum.",
        "links": [
          {
            "description": "Item",
            "url": "https://example.com",
          },
          {
            "description": "Item",
            "url": "https://example.com",
          },
        ],
        "title": "SubCategory with Block Quote and List",
      },
    ],
    "title": "Category with Subcategories",
  },
  {
    "description": "Sed sit amet sollicitudin dolor. Donec scelerisque sed tortor sed elementum.",
    "subcategories": [
      {
        "links": [
          {
            "description": "Item",
            "url": "https://example.com",
          },
          {
            "description": "Item",
            "url": "https://example.com",
          },
        ],
        "subcategories": [
          {
            "description": "Sed sit amet sollicitudin dolor. Donec scelerisque sed tortor sed elementum.",
            "links": [
              {
                "description": "Item",
                "url": "https://example.com",
              },
              {
                "description": "Item",
                "url": "https://example.com",
              },
            ],
            "title": "Deep Category with Block Quote and List",
          },
        ],
        "title": "SubCategory with List",
      },
    ],
    "title": "Category with Block Quote and Subcategories",
  },
]
`;
